[GPU] Use C++11 `std::atomic` instead of non-standard extensions

FabianSchuetze opened the pull request:
As mentioned in #3987 we can use the memory model of c++11 to replace (outdated) cv code. This PR tries to implement the suggestions in the issue. I always find the atomic reference counting a bit strange, but it is implemented the same way in the std::shared_ptr,  and I think it's good to stick with it. 

I am looking forward to any suggestions and comments!

kunaltyagi commented on gpu/containers/src/device_memory.cpp:
```diff
@@ -211,15 +183,15 @@ pcl::gpu::DeviceMemory::DeviceMemory(const DeviceMemory& other_arg)
 {
   if (refcount_)
-    CV_XADD(refcount_, 1);
+    refcount_->fetch_add(1, std::memory_order_seq_cst);
```
Default is seq_cst (strongest). We can skip that I think

kunaltyagi commented on gpu/containers/src/device_memory.cpp:
```diff
@@ -243,9 +215,8 @@ pcl::gpu::DeviceMemory::create(std::size_t sizeBytes_arg)
-    // refcount_ = (int*)cv::fastMalloc(sizeof(*refcount_));
-    refcount_ = new int;
-    *refcount_ = 1;
+    refcount_ = new std::atomic<int>;
```
Thoughts on initializing in single line?

kunaltyagi commented on gpu/containers/src/device_memory.cpp:
```diff
@@ -161,34 +161,6 @@ pcl::gpu::DeviceMemory2D::empty() const
 
 #else
 
-//////////////////////////    XADD    ///////////////////////////////
```
Don't delete. Deprecate instead.

How to deprecate? Use std::atomic in CV_XADD function, deprecate the function  and remove everything else

mvieth replied:
@larshg What's your view on this? Delete or deprecate?

larshg replied:
I'm fine with both. But I guess the safer approach is to deprecate like is done now. However, its probably not used elsewhere.

mvieth replied:
I would generally also prefer the safer approach (deprecating), but in this case, I simply don't see how anyone else could possibly use `CV_XADD`: it is a macro/inline function in a cpp file, so unless someone `#includes` device_memory.cpp in their own code (obviously wrong), there is no way to have access to this?

larshg replied:
Yeah, that's true. I'm also fine with removing it.

kunaltyagi reviewed (commented):
We can also make the default value of refcount 0 to align more with shared_ptr, but not in this PR

FabianSchuetze commented:
Thank you so much, Kunal, for the review! As always, all the comments are very helpful and I will incorporate them!

mvieth commented:
@FabianSchuetze What is the status here? Are you still working on this pull request?

FabianSchuetze commented:
I beg your pardon @mvieth  for the radio silence & thank you for pinging me. I changed jobs and moved. That should not have prevented me from continuing to work on this but it did. I made another commit and look forward to the feedback!

FabianSchuetze commented on gpu/containers/src/device_memory.cpp:
```diff
@@ -36,14 +36,12 @@
 
 #include <pcl/gpu/containers/device_memory.h>
 #include <pcl/gpu/utils/safe_call.hpp>
+#include <pcl/pcl_macros.h>
```
I have tried to add the header, as you suggested, @mvieth . Thanks for pointing this out!

larshg reviewed (approved):
I have run the unit test we have enabled for GPUl, so hopefully they cover the device_memory usage 😄

kunaltyagi reviewed (approved):
> I have run the unit test
> Macro is in cpp file

- [x] Ci is green

No objections from my side

mvieth reviewed (approved):
Thank you @FabianSchuetze
